File: pkg/dslgoja/protobuf_contract_test.go
package dslgoja

import (
	"encoding/json"
	"testing"

	dslv1 "github.com/go-go-golems/hair-booking/gen/proto/fringe/dsl/v1"
	"google.golang.org/protobuf/encoding/protojson"
	"google.golang.org/protobuf/types/known/structpb"
)

func TestFlowStateProtoConversionUsesCamelCaseAndDynamicProps(t *testing.T) {
	result := &InteractionResult{
		SessionID:   "flow_1",
		PageVersion: 3,
		Page: Page{
			SchemaVersion: 1,
			ID:            "intake-service",
			Title:         "Service",
			Shell: Shell{Kind: "intake", Props: map[string]any{
				"step": float64(1),
				"actions": map[string]any{
					"next": map[string]any{"id": "act_next", "event": "next"},
				},
			}},
			Nodes: []Node{{
				Kind: "segmented",
				Props: map[string]any{
					"value": "color",
					"actions": map[string]any{
						"change": map[string]any{"id": "act_category", "event": "change"},
					},
				},
				Meta: &NodeMeta{ID: "category-tabs"},
			}},
		},
	}

	state, err := FlowStateFromResult(result)
	if err != nil {
		t.Fatalf("FlowStateFromResult: %v", err)
	}
	encoded, err := protojson.Marshal(state)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}

	var decoded map[string]any
	if err := json.Unmarshal(encoded, &decoded); err != nil {
		t.Fatalf("json decode: %v", err)
	}
	if decoded["sessionId"] != "flow_1" {
		t.Fatalf("sessionId = %#v", decoded["sessionId"])
	}
	if decoded["pageVersion"] != float64(3) {
		t.Fatalf("pageVersion = %#v", decoded["pageVersion"])
	}
	page := decoded["page"].(map[string]any)
	if page["schemaVersion"] != float64(1) {
		t.Fatalf("schemaVersion = %#v", page["schemaVersion"])
	}
	nodes := page["nodes"].([]any)
	node := nodes[0].(map[string]any)
	meta := node["meta"].(map[string]any)
	if meta["id"] != "category-tabs" {
		t.Fatalf("node meta id = %#v", meta["id"])
	}
	propsJSON := node["props"].(map[string]any)
	if propsJSON["value"] != "color" {
		t.Fatalf("props value = %#v", propsJSON["value"])
	}
}

func TestInteractionEventFromProtoConvertsDynamicValue(t *testing.T) {
	var protoEvent dslv1.InteractionEvent
	if err := protojson.Unmarshal([]byte(`{
		"eventId":"evt_1",
		"sessionId":"flow_1",
		"pageVersion":4,
		"nodeId":"tone-chips",
		"nodeKind":"chipGroup",
		"actionId":"act_tones",
		"event":"change",
		"value":["warm","dimensional"],
		"meta":{"source":"test"}
	}`), &protoEvent); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}

	event := InteractionEventFromProto(&protoEvent)
	if event.EventID != "evt_1" || event.PageVersion != 4 || event.NodeID != "tone-chips" {
		t.Fatalf("event = %#v", event)
	}
	values := event.Value.([]any)
	if values[0] != "warm" || values[1] != "dimensional" {
		t.Fatalf("value = %#v", event.Value)
	}
	if event.Meta["source"] != "test" {
		t.Fatalf("meta = %#v", event.Meta)
	}
}

func TestDSLProtoJSONContractUsesCamelCaseAndDynamicProps(t *testing.T) {
	props, err := structpb.NewStruct(map[string]any{
		"value": "color",
		"actions": map[string]any{
			"change": map[string]any{"id": "act_category", "event": "change"},
		},
	})
	if err != nil {
		t.Fatalf("props struct: %v", err)
	}

	page := &dslv1.Page{
		SchemaVersion: 1,
		Id:            "intake-service",
		Title:         "Service",
		Shell: &dslv1.Shell{
			Kind: "intake",
		},
		Nodes: []*dslv1.Node{{
			Kind:  "segmented",
			Props: props,
			Meta:  &dslv1.NodeMeta{Id: "category-tabs"},
		}},
	}

	encoded, err := protojson.Marshal(page)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}

	var decoded map[string]any
	if err := json.Unmarshal(encoded, &decoded); err != nil {
		t.Fatalf("json decode: %v", err)
	}
	if decoded["schemaVersion"] != float64(1) {
		t.Fatalf("schemaVersion = %#v", decoded["schemaVersion"])
	}
	nodes := decoded["nodes"].([]any)
	node := nodes[0].(map[string]any)
	meta := node["meta"].(map[string]any)
	if meta["id"] != "category-tabs" {
		t.Fatalf("node meta id = %#v", meta["id"])
	}
	propsJSON := node["props"].(map[string]any)
	if propsJSON["value"] != "color" {
		t.Fatalf("props value = %#v", propsJSON["value"])
	}
}

File: web/src/page-dsl/ProtobufContract.test.ts
import { describe, expect, it } from "vitest";
import { fromJson, toJson } from "@bufbuild/protobuf";
import { DslErrorSchema, FlowStateSchema, InteractionEventSchema, PageSchema } from "../pb/proto/fringe/dsl/v1/dsl_pb";

describe("DSL protobuf JSON contract", () => {
  it("decodes protobuf DSL errors", () => {
    const error = fromJson(DslErrorSchema, {
      code: "dsl_session_not_found",
      message: "DSL session not found",
      details: { sessionId: "flow_missing" },
    });

    expect(error.code).toBe("dsl_session_not_found");
    expect(error.message).toBe("DSL session not found");
    expect(error.details).toMatchObject({ sessionId: "flow_missing" });
  });

  it("decodes the central FlowState transport envelope", () => {
    const state = fromJson(FlowStateSchema, {
      sessionId: "flow_1",
      pageVersion: 3,
      page: {
        schemaVersion: 1,
        id: "intake-service",
        title: "Service",
        shell: { kind: "intake", props: { step: 1, total: 7 } },
        nodes: [{
          kind: "summaryRow",
          meta: { id: "estimate-service" },
          props: {
            label: "Service",
            value: "Color",
            actions: { edit: { id: "act_edit", event: "edit" } },
          },
        }],
      },
      effects: [{ kind: "toast", tone: "info", message: "Saved" }],
    });

    expect(state.sessionId).toBe("flow_1");
    expect(state.pageVersion).toBe(3);
    expect(state.page?.nodes[0]?.props?.actions).toMatchObject({ edit: { id: "act_edit" } });
    expect(toJson(FlowStateSchema, state)).toMatchObject({
      sessionId: "flow_1",
      page: { id: "intake-service" },
      effects: [{ kind: "toast", tone: "info", message: "Saved" }],
    });
  });

  it("decodes page JSON with dynamic node props", () => {
    const page = fromJson(PageSchema, {
      schemaVersion: 1,
      id: "intake-service",
      title: "Service",
      shell: { kind: "intake", props: { step: 1, total: 2 } },
      nodes: [{
        kind: "segmented",
        meta: { id: "category-tabs" },
        props: {
          value: "color",
          actions: { change: { id: "act_category", event: "change" } },
        },
      }],
    });

    expect(page.schemaVersion).toBe(1);
    expect(page.nodes[0]?.meta?.id).toBe("category-tabs");
    expect(page.nodes[0]?.props?.value).toBe("color");
  });

  it("encodes interaction events without undefined optional value fields", () => {
    const event = fromJson(InteractionEventSchema, {
      eventId: "evt_edit",
      pageVersion: 5,
      nodeId: "estimate-service",
      nodeKind: "summaryRow",
      actionId: "act_edit",
      event: "edit",
    });

    expect(toJson(InteractionEventSchema, event)).toMatchObject({
      eventId: "evt_edit",
      pageVersion: 5,
      nodeId: "estimate-service",
      event: "edit",
    });
    expect(toJson(InteractionEventSchema, event)).not.toHaveProperty("value");
  });

  it("decodes interaction events with JSON values", () => {
    const event = fromJson(InteractionEventSchema, {
      eventId: "evt_1",
      pageVersion: 3,
      nodeId: "tone-chips",
      nodeKind: "chipGroup",
      actionId: "act_tones",
      event: "change",
      value: ["warm", "dimensional"],
      meta: { source: "test" },
    });

    expect(event.eventId).toBe("evt_1");
    expect(event.pageVersion).toBe(3);
    expect(toJson(InteractionEventSchema, event)).toMatchObject({
      value: ["warm", "dimensional"],
      meta: { source: "test" },
    });
  });
});
